src/main/scala/calc/Compiler.scala

```diff
@@ -62,21 +61,212 @@ object Compiler {
   }
 
   /** Compile an expression tree into an IR `Tree`, which is an expression
-   *  that evaluates to the result of the tree.
-   *
-   *  This is the main method you have to implement.
-   */
-  def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
+    *  that evaluates to the result of the tree.
+    *
+    *  This is the main method you have to implement.
+    */
+  def compileExpr(tree: Tree, etaExpansion: Boolean = true): irt.Tree = {
     implicit val pos = tree.pos
 
-    tree match {
-      case Literal(value) =>
-        irt.DoubleLiteral(value)
+    val foreignEnv = Foreign.staticJavaMethods(
+      clsName = "java.lang.Math",
+      importedMethods = List(
+        ("sin", TFun(1)),
+        ("cos", TFun(1)),
+        ("sqrt", TFun(1)),
+        ("pow", TFun(2))
+      )
+    )
+
+    // Phase 1: Convert AST to Typed AST
+    implicit val typeEnv = Typer.emptyEnv ++ foreignEnv
+    val (typedAst, _) = Typer.inferType(tree)
+
+    // Phase 2: Compile Typed AST to IR
+    val ir = expr(typedAst)
+
+    // Phase 3: Prepend Eta expansion of foreign method to IR
+    if (etaExpansion) {
+      Foreign.prependEtaExpansion(ir, foreignEnv)
```

Comment:
Ideally, you would only prepend the eta-expansion of functions that are actually used in the `typedAst` :-)
Detecting this is similar to identify free variables.